### PR TITLE
Helm: add hpa for read, write and backend

### LINKED ISSUE #7864: Helm chart of loki-read simple scalable lacks HPA (HorizontalPodAutoscaler)
**Describe the bug**
The HPA (HorizontalPodAutoscaler) attributes are mentioned in [values](./production/helm/loki/values.yaml) and [reference](./docs/sources/installation/helm/reference.md) docs, as `read.autoscaling.enabled` for example.

However there's no reference of any HorizontalPodAutoscaler kind nor those attributes in **read** part of chart. The gateway part is has HPA explicitly as an example.

**To Reproduce**
Look at the chart files or try to deploy HPA for **read** Loki pods.

**Expected behavior**
Set HorizontalPodAutoscaler kind or provide an alternative to scale `loki-read` StatefulSet.

Or maybe the HPA was there, bug got removed for some reason and only docs are not being cleaned up?

**Environment:**
 - Infrastructure: Kubernetes, simple-scalable deployment
 - Deployment tool: Helm via FluxCD in my case

**Screenshots, Promtail config, or terminal output**
N/A

### PR DESCRIPTION
**Pull request by oleksii-boiko-ua**
**What this PR does / why we need it**:
It adds missing HPA (HorizontalPodAutoscaler) templates to write, read, and backend components 
**Which issue(s) this PR fixes**:
Fixes #7864

**Checklist**
- [x] Reviewed the [`CONTRIBUTING.md`](https://github.com/grafana/loki/blob/main/CONTRIBUTING.md) guide (**required**)
- [x] Documentation added
- [ ] Tests updated
- [x] `CHANGELOG.md` updated
- [x] Changes that require user attention or interaction to upgrade are documented in `docs/sources/upgrading/_index.md`

### CODE REVIEW

**Comment by jeschkies** [comment]
Thanks. We had plans to add autoscaling but didn't knew when. @trevorwhitney can the backend be scaled? I thought it was stateful. @salvacorts  could you take a look as well?

**Comment by oleksii-boiko-ua** [comment]
BTW QQ, when are we going to have `Three Scalable Targets` supported and merged into the main branch? I saw [this guide](https://github.com/grafana/loki/blob/main/docs/sources/installation/helm/migrate-to-three-scalable-targets/index.md) which  refers to `main-f5fbfab-amd64` docker tag which is 2 month old release

File: production/helm/loki/values.yaml
Comment (JStickler) [inline production/helm/loki/values.yaml:608]:
```suggestion
    # -- Enable autoscaling for the write.
```

File: production/helm/loki/values.yaml
Comment (JStickler) [inline production/helm/loki/values.yaml:610]:
```suggestion
    # -- Minimum autoscaling replicas for the write.
```

File: production/helm/loki/values.yaml
Comment (JStickler) [inline production/helm/loki/values.yaml:612]:
```suggestion
    # -- Maximum autoscaling replicas for the write.
```

File: production/helm/loki/values.yaml
Comment (JStickler) [inline production/helm/loki/values.yaml:614]:
```suggestion
    # -- Target CPU utilization percentage for the write.
```

File: production/helm/loki/values.yaml
Comment (JStickler) [inline production/helm/loki/values.yaml:616]:
```suggestion
    # -- Target memory utilization percentage for the write.
```

File: production/helm/loki/values.yaml
Comment (JStickler) [inline production/helm/loki/values.yaml:825]:
```suggestion
    # -- Enable autoscaling for the backend.
```

File: production/helm/loki/values.yaml
Comment (JStickler) [inline production/helm/loki/values.yaml:827]:
```suggestion
    # -- Minimum autoscaling replicas for the backend.
```

File: production/helm/loki/values.yaml
Comment (JStickler) [inline production/helm/loki/values.yaml:829]:
```suggestion
    # -- Maximum autoscaling replicas for the backend.
```

File: production/helm/loki/values.yaml
Comment (JStickler) [inline production/helm/loki/values.yaml:833]:
```suggestion
    # -- Target memory utilization percentage for the backend.
```

**Review by JStickler — COMMENTED** [review]
[Docs squad]  Added missing punctuation and we've standardized on American spelling in the docs.

File: production/helm/loki/templates/write/hpa.yaml
Comment (kd7lxl) [inline production/helm/loki/templates/write/hpa.yaml:18]: It would be helpful to support [`behavior:`](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/#configurable-scaling-behavior) here, to control how dramatic the scaling is, since the writer is stateful. It could also be used to disable scale down.

**Comment by oleksii-boiko-ua** [comment]
@JStickler @kd7lxl @jeschkies updated pr. thanks

**Comment by oleksii-boiko-ua** [comment]
@kd7lxl should we add behavior to all(read, backend) hpa templates?

**Comment by kd7lxl** [comment]
> @kd7lxl should we add behavior to all(read, backend) hpa templates?

Consistency is the least surprising option, so from that standpoint it would be good to add the same options to all. I don't have an immediate use case for behavior on read, but someone else might.

Thanks for working on this! I needed it today and was about to get started writing it when I found your PR.

File: production/helm/loki/values.yaml
Comment (JStickler) [inline production/helm/loki/values.yaml:614]:
```suggestion
    # -- Target CPU utilisation percentage for the write.
```

**Review by JStickler — APPROVED** [review]
[Docs squad]  One more missing period, but other than that, LGTM

**Comment by oleksii-boiko-ua** [comment]
updated, build failure seems not related

**Comment by oleksii-boiko-ua** [comment]
@trevorwhitney  @jeschkies @salvacorts could anyone review it, please?

**Review by trevorwhitney — CHANGES_REQUESTED** [review]
can you walk me through what it's like to scale the `write` and `backend` pods? I imagine automatically scaling up is easier than going down, due to their stateful nature. I love this for `read`, since that's a deployment now. I'm not against this for `write` and `backend`, but could you provide some documentation on how you would scale those components without any data-loss/downtime?

**Comment by kd7lxl** [comment]
> can you walk me through what it's like to scale the `write` and `backend` pods?

Have you seen the doc? https://github.com/grafana/loki/blob/main/docs/sources/operations/storage/wal.md#how-to-scale-updown

File: production/helm/loki/templates/write/statefulset-write.yaml
Comment (trevorwhitney) [inline production/helm/loki/templates/write/statefulset-write.yaml:27]: typo
```suggestion
    whenDeleted: Delete
```

**Comment by trevorwhitney** [comment]
@kd7lxl I have, but that doesn't really answer my question for how this is handled during autocaling. For exmaple, how do you ensure that ingesters flush to storage when they're scaling down? Forgive me if I missed something, but I don't see this PR modifying the `flush_on_shutdown` config, nor calling the flush endpoint (if that's even possible with HPAs?).

In addition to making sure we programmatically handle this, I also think it's important to call out in the Loki operational documentation that enabling auto-scaling will turn on this behavior and clarify how the WAL is handled, and how PVs and PVCs of scaled down ingesters can be treated.

**Comment by oleksii-boiko-ua** [comment]
Hi there, we could use [lifecycle hooks](https://kubernetes.io/docs/concepts/containers/container-lifecycle-hooks/), for example pre stop, and hit endpoint ingester/flush_shutdown before container goes down, as for pvc it will be cleaned automatically

**Comment by kd7lxl** [comment]
> Hi there, we could use [lifecycle hooks](https://kubernetes.io/docs/concepts/containers/container-lifecycle-hooks/), for example pre stop, and hit endpoint ingester/flush_shutdown before container goes down, as for pvc it will be cleaned automatically

There is a request in #8419 to add the lifecycle hook, like I did for the Cortex helm chart. This isn't a perfect solution because it runs every time the pod stops, even though it's unnecessary for typical pod restarts where the WAL is resumed. It's only useful for scaledown events. If you were to configure your HPA behavior to disable scaledown, the lifecycle hook would not be needed. However, if you want to support scaledown it works great in Cortex.

**Comment by oleksii-boiko-ua** [comment]
you could always have conditional lifecycle hook right?

**Comment by kd7lxl** [comment]
> you could always have conditional lifecycle hook right?

Yes, you could do it like the Cortex chart where it's all in values so the right combination of values overrides will disable scaledown and the lifecycle hook, or you could write an explicit condition (like `.write.autoscaling.scaledown.enabled`) for it that automatically injects all the corresponding settings.

**Comment by oleksii-boiko-ua** [comment]
@trevorwhitney updated

File: production/helm/loki/values.yaml
Comment (trevorwhitney) [inline production/helm/loki/values.yaml:660]: instead of just a comment, how would you feel about taking an opinionated stance here? 

I would like as a default to set the `lifecycle` on line 107 of the write template to have the `preStop` hook, but with an option to override via `values.yaml`. For example, maybe we make the default value here `null`, and have a template that, if `null` and autoscaling, we include the default `preStop`, but if not `null` and autoscaling, we include whatever was provided in the `values.yaml`.

similarly, if not `null` and not autoscaling, we can include whatever was provided, but if `null` and not autoscaling we don't include anything?

File: production/helm/loki/values.yaml
Comment (trevorwhitney) [inline production/helm/loki/values.yaml:780]: again with behavior, how would you feel about taking a more opinionated stance with good defaults and the ability to override?

**Review by trevorwhitney — CHANGES_REQUESTED** [review]
This looks great, thank you for making those changes! I'm much happier with this approach using the `preStop` hook and really appreciate the added documentation, thank you!

My last little nit is I'd like to provide more opinionated defaults. See my comments, but I'd like to inject more logic to automatically set the `lifecycle` and `behavior` values to good defaults when using autoscaling rather than hiding best practices in a comment. Of course we need to preserve the ability to change them or not set them at all, but if someone turns on "yes I want autoscaling" let's give them sensible defaults. WDYT?

File: production/helm/loki/values.yaml
Comment (oleksii-boiko-ua) [reply production/helm/loki/values.yaml:660]: sure, updated

File: production/helm/loki/values.yaml
Comment (oleksii-boiko-ua) [reply production/helm/loki/values.yaml:780]: TBH, while the default configuration for the write component is conservative to avoid frequent scale-down events, it's difficult to establish optimal default settings for other components, as their ideal scaling behavior depends on the specific environment and workload.

**Comment by oleksii-boiko-ua** [comment]
@trevorwhitney updated

File: production/helm/loki/values.yaml
Comment (trevorwhitney) [reply production/helm/loki/values.yaml:781]: yeah, fair point, agreed. thanks for the feedback!

**Review by trevorwhitney — APPROVED** [review]
Thanks so much for all you effort on this, looks great! Just a few build failures to get straightened out.

**Comment by oleksii-boiko-ua** [comment]
@kd7lxl could you take a look?

File: docs/sources/operations/storage/wal.md
Comment (kd7lxl) [inline docs/sources/operations/storage/wal.md:104]: This is a little misleading. The `terminationGracePeriodSeconds` is what provides time for the ingester to flush its data before being deleted.

https://github.com/grafana/loki/blob/425736614c43f3435e52bc1e5a2295dfcbb6418a/production/helm/loki/values.yaml#L679-L682

If you think flushing data may take 30 minutes, you might recommend increasing the `terminationGracePeriodSeconds`.

The scaledown policy is more important to avoid thrashing and for queries when configured to only query the ingesters. If all replicas of a stream are flushed from ingesters due to multiple pods scaledown within the `max_chunk_age`, the query will rely on the store to return the data. My understanding is that default Loki configuration handles this gracefully, so it would only be a concern with an odd config like `query_ingester_only: true`.

File: docs/sources/operations/storage/wal.md
Comment (kd7lxl) [inline docs/sources/operations/storage/wal.md:100]: You describe the behavior of `OrderedReady` here, but the statefulset is configured for `Parallel`.

https://github.com/grafana/loki/blob/425736614c43f3435e52bc1e5a2295dfcbb6418a/production/helm/loki/templates/write/statefulset-write.yaml#L15

File: production/helm/loki/templates/write/statefulset-write.yaml
Comment (kd7lxl) [inline production/helm/loki/templates/write/statefulset-write.yaml:109]: Not a big deal, but this could be a single `else if`.
```suggestion
          {{- else if .Values.write.autoscaling.enabled }}
```

**Review by kd7lxl — COMMENTED** [review]
@oleksii-boiko-ua, per your request, I've provided a review. Note that I am not a maintainer of this project.

The way you've organized the values is simple to use and easy to override. I like it. Just made a few comments on wording in the docs.

File: docs/sources/operations/storage/wal.md
Comment (oleksii-boiko-ua) [reply docs/sources/operations/storage/wal.md:100]: Updated, thanos good catch

File: docs/sources/operations/storage/wal.md
Comment (oleksii-boiko-ua) [reply docs/sources/operations/storage/wal.md:102]: fixed

**Comment by oleksii-boiko-ua** [comment]
@kd7lxl updated. thanks

**Comment by oleksii-boiko-ua** [comment]
@trevorwhitney @jeschkies could anyone merge it, or what is the process?

**Comment by aglees** [comment]
Is there any news on when this might be released?

**Comment by HerbBoy** [comment]
Is there a policy preventing this from being merged? Or when can we expect this fix to be implemented?

**Comment by MichelHollands** [comment]
@oleksii-boiko-ua Can you add a HELM changelog entry and update the Helm chart version to 5.5.6? An example can be found [here](https://github.com/grafana/loki/commit/d10549e3ece02120974929894ee333d07755d213). 
Also please rebase and fix the 2 small CI issues.
Sorry for the delay. There are a lot of Helm PRs coming in.

**Comment by oleksii-boiko-ua** [comment]
@MichelHollands updated. thanks

**Review by MichelHollands — APPROVED** [review]
LGTM

**Comment by MichelHollands** [comment]
@oleksii-boiko-ua There is a parse error when the chart is installed. See the failing CI step. Can you have a look at that?

**Pull request merged**
